src/checkers/inference/InferenceQualifierHierarchy.java

```diff
@@ -224,9 +224,15 @@ public boolean isSubtype(final AnnotationMirror subtype, final AnnotationMirror
 
         final Slot subSlot   = slotMgr.getSlot(subtype);
         final Slot superSlot = slotMgr.getSlot(supertype);
-        constraintMgr.addSubtypeConstraint(subSlot, superSlot);
-
-        return true;
+        if (subSlot instanceof ConstantSlot && superSlot instanceof ConstantSlot) {
```

Comment:
I think one concern I saw with previous reviews was that this duplicates logic, possible in multiple places.
I'm wondering whether we can design a class that encapsulates this logic, by taking the constraint type and slots, and then returning: TRUE, FALSE, UNKNOWN, REPLACE_WITH_EQUALITY, etc.

For Subtype Constraints, this would be:

```
C1 <: C2 => TRUE/FALSE depending on relationship
V1 <: V2 => UNKNOWN
V1 <: TOP => TRUE
BOTTOM <: V2 => TRUE
TOP <: V2 => REPLACE_WITH_EQUALITY
V2 <: BOTTOM => REPLACE_WITH_EQUALITY
```

Could something like this be used here and in ConstraintManager?
How about the other constraint classes?

Reply:
> `I think one concern I saw with previous reviews was that this duplicates logic, possible in multiple places.`

I don't think there are other duplicates except in overriding `leastUpperBound()` in `InferenceQualifierHierarchy`. Can you provide one example for duplicate?

> `I'm wondering whether we can design a class that encapsulates this logic, by taking the constraint type and slots, and then returning: TRUE, FALSE, UNKNOWN, REPLACE_WITH_EQUALITY, etc.`

I think this is over-complicated refactoring. We still need to find all the places that have "duplicate logic" and insert the method calls to the newly added class in every occurence. But I didn't find there are duplicates other than `isSubtype()` and `leastUpperBound()`.

Reply:
> I think this is over-complicated refactoring.
Pre-processing the cases would make inference performance faster.

Reply:
Ok, if there is no duplication of this logic, I'm fine with coding it here.
In general, we need to think about whether the InferenceQualifierHierarchy is used consistently, that is, only used to enforce a subtype relationship. General tests that do different things when true/false gets returned don't work correctly.